### PR TITLE
fabric: add fabric-gateway-controller deployment

### PR DESCRIPTION
**Pull request by AlexanderYastrebov**
Signed-off-by: Alexander Yastrebov <alexander.yastrebov@zalando.de>

### CODE REVIEW

File: cluster/manifests/deletions.yaml [hunk 1]
```diff
@@ -40,6 +40,16 @@ pre_apply:
...
   kind: Deployment
 
+{{ if eq .ConfigItems.fabric_gateway_controller_mode "off" }}
+- name: fabric-gateway-controller
+  namespace: kube-system
+  kind: Deployment
+- labels:
+    fabric-gateway: "true"
+  kind: RouteGroup
+{{ end }}
```
Comment (AlexanderYastrebov) [inline cluster/manifests/deletions.yaml]: Should this include RBAC deletions as well? The problem is that it takes some time to scale down deployment on deletion which prevents deletion of a ServiceAccount that is in use.

File: cluster/manifests/deletions.yaml [hunk 2]
```diff
@@ -40,6 +40,16 @@ pre_apply:
   namespace: kube-system
   kind: Deployment
 
+{{ if eq .ConfigItems.fabric_gateway_controller_mode "off" }}
+- name: fabric-gateway-controller
+  namespace: kube-system
+  kind: Deployment
+- labels:
+    fabric-gateway: "true"
```
Comment (AlexanderYastrebov) [inline cluster/manifests/deletions.yaml]: Should this label be more "unique"?

File: cluster/manifests/deletions.yaml (hunk 1 above)
Comment (AlexanderYastrebov) [reply cluster/manifests/deletions.yaml]: Decided to avoid deletions altogether

File: cluster/manifests/fabric-gateway/deployment.yaml [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+        application: kubernetes
+        component: fabric-gateway-controller
+        version: {{ $version }}
+      annotations:
+        logging/destination: "{{ .Cluster.ConfigItems.log_destination_infra }}"
+        prometheus.io/path: /metrics
+        prometheus.io/port: "7979"
+        prometheus.io/scrape: "true"
+    spec:
+      serviceAccountName: fabric-gateway-controller
```
Comment (AlexanderYastrebov) [inline cluster/manifests/fabric-gateway/deployment.yaml:37]: Do we need
```yaml
      priorityClassName: "{{ .Cluster.ConfigItems.system_priority_class }}"
```
like in other controllers?

File: cluster/manifests/fabric-gateway/01-rbac.yaml [hunk 4]
```diff
@@ -0,0 +1,72 @@
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  name: fabric-gateway-controller
+  namespace: kube-system
+  labels:
+    application: kubernetes
```
Comment (szuecs) [inline cluster/manifests/fabric-gateway/01-rbac.yaml]: Did you already discuss it with Mikkel?

Comment (szuecs) [reply cluster/manifests/fabric-gateway/01-rbac.yaml]: sounds good to me, not sure about others

Comment (AlexanderYastrebov) [reply cluster/manifests/fabric-gateway/01-rbac.yaml]: I've borrowed deployment structure from other controllers like stackset-controller or external-dns controller

File: cluster/manifests/fabric-gateway/deployment.yaml (hunk 3 above)
Comment (szuecs) [reply cluster/manifests/fabric-gateway/deployment.yaml:37]: yes

File: cluster/manifests/deletions.yaml (hunk 2 above)
Comment (AlexanderYastrebov) [reply cluster/manifests/deletions.yaml]: Decided to avoid deletions altogether

File: cluster/manifests/fabric-gateway/deployment.yaml (hunk 3 above)
Comment (AlexanderYastrebov) [reply cluster/manifests/fabric-gateway/deployment.yaml:37]: Added

File: cluster/manifests/fabric-gateway/01-rbac.yaml (hunk 4 above)
Comment (szuecs) [reply cluster/manifests/fabric-gateway/01-rbac.yaml]: Maybe better to use it similar to kube-ingress-aws-controller

Comment (AlexanderYastrebov) [reply cluster/manifests/fabric-gateway/01-rbac.yaml]: I've changed application label to `gateway-operator` for now. It should be easy to update labels later if needed.

**Comment by szuecs** [comment]
:+1:

**Comment by szuecs** [comment]
:+1:

**Comment by AlexanderYastrebov** [comment]
:+1:

**Pull request merged**